PSPs/drafts/psp-34.md

```diff
@@ -0,0 +1,914 @@
+```json
+{
+  "args": [
+    {
+      "label": "operator",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "label": "from",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "label": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    },
+    {
+      "label": "data",
+      "type": {
+        "displayName": [
+          "[u8]"
+        ],
+        "type": "[u8]"
+      }
+    }
+  ],
+  "docs": [
+    "Ensures that the smart contract allows reception of PSP34 token(s).",
+    "Returns `Ok(())` if the contract allows the reception of the token(s) and Error `TransferRejected(String)` otherwise.",
+    "",
+    "This method will get called on every transfer to check whether the recipient in `transfer`",
+    "or `transfer_from` is a contract, and if it is, does it accept tokens.",
+    "This is done to prevent contracts from locking tokens forever.",
+    "",
+    "Returns `PSP34ReceiverError` if the contract does not accept the tokens."
+  ],
+  "mutates": true,
+  "label": "PSP34Receiver::before_received",
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Result"
+    ],
+    "type": 2
+  },
+  "selector": "0xbb7df780"
+}
+```
+
+### Extension
+
+#### PSP34Metadata
+
+`PSP34Metadata` is a **recommended** extension for this Non-Fungible Token standard
+because it is main feature of the NFT.
+
+##### **total_supply**() ➔ Balance
+Selector: `0x50a9188c` - first 4 bytes of `blake2b_256("PSP34Metadata::total_supply")`
+```json
+{
+  "args": [],
+  "docs": [
+    "Returns the current total supply of the NFT."
+  ],
+  "mutates": false,
+  "label": "PSP34Metadata::total_supply",
+  "returnType": {
+    "displayName": [
+      "Balance"
+    ],
+    "type": "Balance"
+  },
+  "selector": "0x50a9188c"
+}
+```
+
+##### **get_attribute**(id: Id, key: [u8]) ➔ Option<[u8]>
+Selector: `0xf19d48d1` - first 4 bytes of `blake2b_256("PSP34Metadata::get_attribute")`
+```json
+{
+  "args": [
+    {
+      "label": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    },
+    {
+      "label": "key",
+      "type": {
+        "displayName": [
+          "[u8]"
+        ],
+        "type": "[u8]"
+      }
+    }
+  ],
+  "docs": [
+    "Returns the attribute of `id` for the given `key`.",
+    "",
+    "If `id` is a collection id of the token, it returns attributes for collection."
+  ],
+  "mutates": false,
+  "label": "PSP34Metadata::get_attribute",
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Option"
+    ],
+    "type": "Option<[u8]>"
+  },
+  "selector": "0xf19d48d1"
+}
+```
+
+The list of required attributes for NFT should be defined in a separate 
+proposal based on the scope of the usage.
+
+### Events
+
```

Comment:
Do you think that we need to provide event signatures as well?

Reply:
Yep, because the signature of the event defines topics of that event + it is used during decoding of event on indexer side

Reply:
Let's add it then.

Reply:
I meant the ABI already is a signature=D How it is defined in the Rust is not important

Reply:
At least name is important, as it is used for the label.